src/main/java/com/igot/cb/access_settings/service/impl/CourseServiceImpl.java

```diff
@@ -110,9 +133,217 @@ public ApiResponse readContentState(Map<String, Object> requestBody, String auth
+            } else if (existingAccessTimeObj instanceof Date) {
+                existingAccessTime = (Date) existingAccessTimeObj;
+            } else {
+                existingAccessTime = null;
+            }
+            if (existingAccessTime == null) {
+                existingAccessTime = parseDate((String) existingContent.getOrDefault(Constants.OLD_LAST_ACCESS_TIME, ""));
+            }
+            updatedContent.put(Constants.LAST_ACCESS_TIME, compareTime(existingAccessTime, inputAccessTime));
+
+            int inputProgress = ((Number) inputContent.getOrDefault(Constants.PROGRESS, 0)).intValue();
+            int existingProgress = ((Number) existingContent.getOrDefault(Constants.PROGRESS, 0)).intValue();
+            updatedContent.put(Constants.PROGRESS, Math.max(inputProgress, existingProgress));
+
+            int existingStatus = ((Number) existingContent.getOrDefault(Constants.STATUS, 0)).intValue();
+            Object existingCompletedTimeObj = existingContent.get(Constants.LAST_COMPLETED_TIME);
+            Date existingCompletedTime;
+            if (existingCompletedTimeObj instanceof String) {
+                existingCompletedTime = parseDate((String) existingCompletedTimeObj);
+            } else if (existingCompletedTimeObj instanceof Date) {
+                existingCompletedTime = (Date) existingCompletedTimeObj;
+            } else {
+                existingCompletedTime = null;
+            }
+            if (existingCompletedTime == null) {
+                existingCompletedTime = parseDate((String) existingContent.getOrDefault(Constants.OLD_LAST_COMPLETED_TIME, ""));
+            }
+
+            if (inputStatus >= existingStatus) {
+                if (inputStatus >= 2) {
+                    updatedContent.put(Constants.STATUS, 2);
+                    updatedContent.put(Constants.PROGRESS, 100);
+                    updatedContent.put(Constants.LAST_COMPLETED_TIME, compareTime(existingCompletedTime, inputCompletedTime));
+                }
+            } else {
+                updatedContent.put(Constants.STATUS, existingStatus);
+            }
+        } else {
+            if (inputStatus >= 2) {
+                updatedContent.put(Constants.PROGRESS, 100);
+                updatedContent.put(Constants.LAST_COMPLETED_TIME, compareTime(null, inputCompletedTime));
+                updatedContent.put(Constants.STATUS, 2);
+            } else {
+                updatedContent.put(Constants.PROGRESS, 0);
+            }
+            updatedContent.put(Constants.LAST_ACCESS_TIME, compareTime(null, inputAccessTime));
+        }
+
+        updatedContent.put(Constants.LAST_UPDATED_TIME, Instant.now());
+        updatedContent.put(Constants.USER_ID, userId);
+        updatedContent.replaceAll((k, v) -> v instanceof Timestamp ? ((Timestamp) v).toInstant() : v);
+
+        return updatedContent;
+    }
+    public Date parseDate(String dateString) {
```

Comment:
Is there any time zone is used in existing sunbird code?
will this code guarantee the time from client would be same as server ?